// protopedia-api-v2-client/lib/src/client.js
import { createTimeoutError, headersToPlainObject, isAbortError, mergeHeaders, serializeListPrototypeParams, toTimeout, trimTrailingSlash, } from './client-utils.js';
import { ProtoPediaApiError } from './errors.js';
import { createLoggerConfig, getLoggerMethod, getLogLevelValue, headersForLogging, shouldLog, } from './logger.js';
import { VERSION } from './version.js';
const DEFAULT_USER_AGENT = `ProtoPedia API Ver 2.0 Node.js Client/${VERSION}`;
const DEFAULT_BASE_URL = 'https://protopedia.net/v2/api';
const DEFAULT_TIMEOUT_MS = 15_000;
export class ProtoPediaApiClient {
    token;
    baseUrl;
    fetchFn;
    timeoutMs;
    userAgent;
    logger;
    logLevel;
    logLevelValue;
    /**
     * Creates a client capable of interacting with the ProtoPedia v2 API.
     *
     * @param options Optional configuration such as authentication, base URL,
     * and logging preferences.
     */
    constructor(options = {}) {
        const fetchFn = options.fetch ?? globalThis.fetch;
        if (typeof fetchFn !== 'function') {
            throw new Error('fetch is not available. Provide a fetch implementation in options.fetch.');
        }
        this.fetchFn = fetchFn;
        this.token = options.token;
        this.baseUrl = trimTrailingSlash(options.baseUrl ?? DEFAULT_BASE_URL);
        this.timeoutMs = toTimeout(options.timeoutMs, DEFAULT_TIMEOUT_MS);
        this.userAgent = options.userAgent ?? DEFAULT_USER_AGENT;
        const loggerConfig = createLoggerConfig(options.logger, options.logLevel);
        this.logger = loggerConfig.logger;
        this.logLevel = loggerConfig.level;
        this.logLevelValue = loggerConfig.levelValue;
    }
    /**
     * Retrieves a paginated list of ProtoPedia prototypes.
     *
     * @see https://protopediav2.docs.apiary.io/#reference/0/0/0?console=1
     */
    async listPrototypes(params, requestOptions) {
        const logLevelValue = this.resolveLogLevelValue(requestOptions);
        const url = this.buildUrl('/prototype/list', serializeListPrototypeParams(params));
        const method = 'GET';
        const response = await this.execute(url, {
            method,
            headers: {
                Accept: 'application/json',
            },
        }, requestOptions, logLevelValue);
        const payload = (await parseJson(response, method, 'listPrototype'));
        this.log('debug', 'listPrototype response payload', payload, logLevelValue);
        return payload;
    }
    /**
     * Downloads prototype metadata as a TSV document.
     *
     * @see https://protopediav2.docs.apiary.io/#reference/0/(tsv)/0
     */
    async downloadPrototypesTsv(params, requestOptions) {
        const logLevelValue = this.resolveLogLevelValue(requestOptions);
        const url = this.buildUrl('/prototype/list/tsv', serializeListPrototypeParams(params));
        const method = 'GET';
        const response = await this.execute(url, {
            method,
            headers: {
                Accept: 'application/json',
            },
        }, requestOptions, logLevelValue);
        const text = await response.text();
        this.log('debug', 'downloadPrototypesTsv response payload', text, logLevelValue);
        return text;
    }
    /**
     * Performs the underlying HTTP request and wraps common logging and error handling.
     */
    async execute(url, options, requestOptions, configuredLevelValue) {
        const logLevelValue = configuredLevelValue ?? this.resolveLogLevelValue(requestOptions);
        const { method, body } = options;
        const headers = mergeHeaders(this.token ? { Authorization: `Bearer ${this.token}` } : undefined, this.userAgent ? { 'X-Client-User-Agent': this.userAgent } : undefined, options.headers, requestOptions?.headers);
        const { signal, cleanup } = this.createAbortSignal(requestOptions?.signal);
        const init = {
            method,
            headers,
            signal,
        };
        if (body !== undefined) {
            init.body = body;
        }
        try {
            this.log('debug', 'HTTP request', {
                method,
                url,
                headers: headersForLogging(headers),
            }, logLevelValue);
            const response = await this.fetchFn(url, init);
            this.log('debug', 'HTTP response', {
                method,
                url,
                status: response.status,
                headers: headersForLogging(response.headers),
            }, logLevelValue);
            if (!response.ok) {
                const apiError = await buildError(response, method);
                this.log('error', 'API request failed', apiError, logLevelValue);
                throw apiError;
            }
            return response;
        }
        catch (error) {
            if (error instanceof ProtoPediaApiError) {
                throw error;
            }
            if (isAbortError(error) && requestOptions?.signal?.aborted) {
                this.log('warn', 'HTTP request aborted by caller', {
                    method,
                    url,
                    reason: requestOptions.signal.reason,
                }, logLevelValue);
                throw requestOptions.signal.reason ?? error;
            }
            if (isAbortError(error)) {
                this.log('warn', 'HTTP request aborted', {
                    method,
                    url,
                    error,
                }, logLevelValue);
            }
            else {
                this.log('error', 'HTTP request threw an unexpected error', {
                    method,
                    url,
                    error,
                }, logLevelValue);
            }
            throw error;
        }
        finally {
            cleanup();
        }
    }
    buildUrl(pathname, query) {
        const normalizedPath = pathname.replace(/^\/+/, '');
        const base = `${this.baseUrl}/`;
        const url = new URL(normalizedPath, base);
        query.forEach((value, key) => {
            if (value !== '') {
                url.searchParams.set(key, value);
            }
        });
        return url.toString();
    }
    createAbortSignal(signal) {
        const controller = new AbortController();
        const timeoutMs = this.timeoutMs;
        let timeoutId;
        let abortListener;
        if (Number.isFinite(timeoutMs) && timeoutMs > 0) {
            timeoutId = setTimeout(() => {
                controller.abort(createTimeoutError(timeoutMs));
            }, timeoutMs);
        }
        if (signal) {
            if (signal.aborted) {
                controller.abort(signal.reason);
            }
            else {
                abortListener = () => {
                    controller.abort(signal.reason);
                };
                signal.addEventListener('abort', abortListener, { once: true });
            }
        }
        const cleanup = () => {
            if (timeoutId) {
                clearTimeout(timeoutId);
            }
            if (signal && abortListener) {
                signal.removeEventListener('abort', abortListener);
            }
        };
        return { signal: controller.signal, cleanup };
    }
    log(level, message, metadata, configuredLevelValue) {
        const levelValue = configuredLevelValue ?? this.logLevelValue;
        if (!shouldLog(levelValue, level)) {
            return;
        }
        const method = getLoggerMethod(this.logger, level);
        if (metadata === undefined) {
            method(message);
        }
        else {
            method(message, metadata);
        }
    }
    resolveLogLevelValue(requestOptions) {
        if (!requestOptions?.logLevel) {
            return this.logLevelValue;
        }
        return getLogLevelValue(requestOptions.logLevel);
    }
}
/**
 * Builds a ProtoPedia API client.
 */
export function createProtoPediaClient(options = {}) {
    const token = options.token;
    if (token == null || !token) {
        throw new Error('Missing PROTOPEDIA_API_V2_TOKEN.');
    }
    return new ProtoPediaApiClient(options);
}
async function parseJson(response, method, context) {
    try {
        return await response.json();
    }
    catch (error) {
        const text = await response.text();
        throw new ProtoPediaApiError({
            message: `Failed to parse ${context} response as JSON`,
            req: {
                method,
                url: response.url,
            },
            status: response.status,
            statusText: response.statusText,
            headers: headersToPlainObject(response.headers),
            body: text,
            cause: error,
        });
    }
}
async function buildError(response, method) {
    let body = null;
    // Prefer JSON when content-type hints JSON; otherwise text. Use clone() to
    // avoid consuming the body twice and causing "Body has already been read".
    const contentType = response.headers.get('content-type')?.toLowerCase() ?? '';
    if (contentType.includes('application/json')) {
        try {
            const copy = response.clone();
            body = await copy.json();
        }
        catch {
            try {
                body = await response.text();
            }
            catch (readError) {
                body = { parseError: readError.message };
            }
        }
    }
    else {
        try {
            body = await response.text();
        }
        catch (textError) {
            try {
                // As a last resort, try JSON on the original response without cloning
                // to support minimal mock Response objects in tests.
                body = await response.json();
            }
            catch {
                body = { parseError: textError.message };
            }
        }
    }
    return new ProtoPediaApiError({
        message: 'API request failed',
        req: {
            method,
            url: response.url,
        },
        status: response.status,
        statusText: response.statusText,
        headers: headersToPlainObject(response.headers),
        body,
    });
}
