// protopedia-api-v2-client/lib/src/errors.js
/**
 * Error type used to represent unsuccessful ProtoPedia API responses.
 */
export class ProtoPediaApiError extends Error {
    req;
    status;
    statusText;
    body;
    headers;
    /**
     * Creates a typed error that mirrors the failing HTTP response produced by the
     * ProtoPedia API.
     */
    constructor(options) {
        super(options.message, { cause: options.cause });
        this.name = 'ProtoPediaApiError';
        this.req = { ...options.req };
        this.status = options.status;
        this.statusText = options.statusText;
        this.body = options.body ?? null;
        this.headers = { ...options.headers };
    }
    /**
     * Converts the error into a JSON serialisable representation suitable for
     * logging or rethrowing across process boundaries.
     */
    toJSON() {
        return {
            name: this.name,
            message: this.message,
            req: this.req,
            status: this.status,
            statusText: this.statusText,
            body: this.body,
            headers: this.headers,
        };
    }
}
