// protopedia-api-v2-client/lib/src/logger.js
/**
 * Mapping of log levels to numeric values for comparison purposes.
 */
const LOG_LEVEL_VALUES = {
    error: 0,
    warn: 1,
    info: 2,
    debug: 3,
};
/**
 * No-operation logger that ignores all messages.
 */
const NOOP_LOGGER = {
    error: () => { },
    warn: () => { },
    info: () => { },
    debug: () => { },
};
/**
 * Normalises logger configuration, providing defaults for missing pieces.
 */
export function createLoggerConfig(logger, level) {
    const resolvedLevel = level ?? 'error';
    const sanitizedLogger = sanitizeLogger(logger ?? createConsoleLogger());
    return {
        logger: sanitizedLogger,
        level: resolvedLevel,
        levelValue: getLogLevelValue(resolvedLevel),
    };
}
/**
 * Determines whether a message with the provided level should be emitted.
 */
export function shouldLog(configuredLevelValue, messageLevel) {
    if (configuredLevelValue < 0) {
        return false;
    }
    const messageLevelValue = getLogLevelValue(messageLevel);
    if (messageLevelValue < 0) {
        return false;
    }
    return messageLevelValue <= configuredLevelValue;
}
/**
 * Converts arbitrary string input into a supported log level, if possible.
 */
export function normaliseLogLevel(input) {
    if (typeof input !== 'string') {
        return undefined;
    }
    const value = input.trim().toLowerCase();
    if (!value) {
        return undefined;
    }
    switch (value) {
        case 'silent':
        case 'error':
        case 'warn':
        case 'warning':
            return value === 'warning' ? 'warn' : value;
        case 'info':
        case 'debug':
            return value;
        default:
            return undefined;
    }
}
/**
 * Provides the numeric representation for a log level, with silent mapped to -1.
 */
export function getLogLevelValue(level) {
    if (level === 'silent') {
        return -1;
    }
    return LOG_LEVEL_VALUES[level];
}
function sanitizeLogger(logger) {
    return {
        error: coalesceLoggerMethod(logger.error),
        warn: coalesceLoggerMethod(logger.warn, logger.error),
        info: coalesceLoggerMethod(logger.info, logger.warn ?? logger.error),
        debug: coalesceLoggerMethod(logger.debug, logger.info ?? logger.warn ?? logger.error),
    };
}
function coalesceLoggerMethod(candidate, fallback) {
    if (typeof candidate === 'function') {
        return wrapMethod(candidate);
    }
    if (typeof fallback === 'function') {
        return wrapMethod(fallback);
    }
    return NOOP_LOGGER.debug;
}
function wrapMethod(method) {
    return (message, metadata) => {
        if (metadata === undefined) {
            method(message);
        }
        else {
            method(message, metadata);
        }
    };
}
function createConsoleLogger() {
    if (typeof console !== 'object' || console === null) {
        return NOOP_LOGGER;
    }
    const fallback = wrapMethod(() => { });
    const error = typeof console.error === 'function'
        ? wrapMethod(console.error.bind(console))
        : fallback;
    const warn = typeof console.warn === 'function'
        ? wrapMethod(console.warn.bind(console))
        : error;
    const info = typeof console.info === 'function'
        ? wrapMethod(console.info.bind(console))
        : warn;
    const debug = typeof console.debug === 'function'
        ? wrapMethod(console.debug.bind(console))
        : info;
    return {
        error,
        warn,
        info,
        debug,
    };
}
/**
 * Selects the method on the logger that corresponds to the desired level.
 */
export function getLoggerMethod(logger, level) {
    return logger[level];
}
/**
 * Converts Headers into a plain object suitable for structured logging.
 *
 * When a header key includes 'token' or 'auth' (case-insensitive),
 * its value is masked as '***' to prevent leaking sensitive data in logs.
 */
export function headersForLogging(headers) {
    if (!headers) {
        return undefined;
    }
    if (headers instanceof Headers) {
        return headersToMaskedObject(headers);
    }
    const constructed = new Headers(headers);
    return headersToMaskedObject(constructed);
}
/**
 * Converts Headers into a plain object suitable for structured logging.
 *
 * When a header key includes 'token' or 'auth' (case-insensitive),
 * its value is masked as '***'.
 *
 * @param headers
 * @returns
 */
export function headersToMaskedObject(headers) {
    const out = {};
    headers.forEach((value, key) => {
        const keyLower = key.toLowerCase();
        const shouldMask = keyLower.includes('token') || keyLower.includes('auth');
        out[key] = shouldMask ? '***' : value;
    });
    return out;
}
